File: src/hetzner_code/commands/create.py
from __future__ import annotations

import tempfile
from datetime import UTC, datetime
from pathlib import Path

import click

from hetzner_code import attach, hetzner, ssh_util, state
from hetzner_code.naming import generate_instance_name
from hetzner_code.provision import BASE_PROVISION_SCRIPT
from hetzner_code.run import HcodeError


def create(
    *,
    repo_url: str,
    name: str | None,
    branch: str | None,
    server_type: str,
    location: str,
    login_key: str | None,
    login_key_path: str,
    env_files: tuple[str, ...],
    no_attach: bool,
) -> None:
    from hetzner_code.github import parse_repo_url

    repo = parse_repo_url(repo_url)
    instance_name = name or generate_instance_name(repo.name)
    if state.exists(instance_name):
        raise HcodeError(
            f"instance '{instance_name}' already exists — pick another --name"
        )

    resolved_login_key = login_key or hetzner.pick_login_key()
    identity_path = Path(login_key_path).expanduser()
    if not identity_path.exists():
        raise HcodeError(f"--login-key-path {identity_path} does not exist")

    click.echo(
        f"[1/5] checking {identity_path} matches '{resolved_login_key}' on Hetzner ..."
    )
    hetzner.verify_login_key(resolved_login_key, identity_path)

    click.echo(f"[2/5] creating {server_type} in {location} ({instance_name}) ...")
    with tempfile.NamedTemporaryFile("w", suffix=".sh", delete=False) as f:
        f.write(BASE_PROVISION_SCRIPT)
        script_path = Path(f.name)
    try:
        server_id, ip = hetzner.create_server(
            name=instance_name,
            server_type=server_type,
            location=location,
            login_key=resolved_login_key,
            user_data_file=script_path,
            labels={"hcode-instance": instance_name},
        )
    finally:
        script_path.unlink(missing_ok=True)

    click.echo(
        f"[3/5] waiting for {ip} to finish booting (this installs Docker/Node/uv, ~1-2 min) ..."
    )
    ssh_util.wait_for_ssh(ip, identity_path)

    click.echo("[4/5] attaching the repo ...")
    repo_state = attach.attach_repo(
        ip=ip,
        identity_path=identity_path,
        instance_name=instance_name,
        repo_url=repo_url,
        branch=branch,
        env_files=env_files,
    )

    instance = state.Instance(
        name=instance_name,
        server_id=server_id,
        ip=ip,
        type=server_type,
        location=location,
        login_key=resolved_login_key,
        login_key_path=str(identity_path),
        created_at=datetime.now(UTC).isoformat(),
        repos=[repo_state],
    )
    state.save(instance)

    click.echo(f"[5/5] {instance_name} is up at {ip}")
    click.echo(f"  ssh:    hcode ssh {instance_name}")
    click.echo(f"  add:    hcode add <repo-url> --instance {instance_name}")
    click.echo(f"  kill:   hcode destroy {instance_name}")

    if not no_attach:
        ssh_util.interactive(ip, identity_path, cwd=f"/root/code/{repo.name}")

File: internal/commands/create.go
package commands

import (
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/dikee/hcode/internal/attach"
	"github.com/dikee/hcode/internal/config"
	"github.com/dikee/hcode/internal/forwards"
	"github.com/dikee/hcode/internal/github"
	"github.com/dikee/hcode/internal/hetzner"
	"github.com/dikee/hcode/internal/naming"
	"github.com/dikee/hcode/internal/provision"
	"github.com/dikee/hcode/internal/run"
	"github.com/dikee/hcode/internal/sshutil"
	"github.com/dikee/hcode/internal/state"
)

// CreateOptions are every flag `hcode create` accepts.
type CreateOptions struct {
	RepoURL      string
	Name         string
	Branch       string
	ServerType   string
	Location     string
	LoginKey     string
	LoginKeyPath string
	EnvFiles     []string
	Worktrees    int
	OpsDir       string
	PostClone    string
	PostWorktree string
	Forwards     []string
	NoAttach     bool
}

// Create runs `hcode create`.
func Create(opts CreateOptions) error {
	repo, err := github.ParseRepoURL(opts.RepoURL)
	if err != nil {
		return err
	}
	instanceName := opts.Name
	if instanceName == "" {
		instanceName = naming.GenerateInstanceName(repo.Name)
	}
	if state.Exists(instanceName) {
		return run.Errorf("instance '%s' already exists — pick another --name", instanceName)
	}

	if len(opts.Forwards) > 0 && opts.NoAttach {
		fmt.Println("note: --forward has no effect with --no-attach — nothing stays open to tunnel through")
	}

	resolvedLoginKey := opts.LoginKey
	if resolvedLoginKey == "" {
		resolvedLoginKey, err = hetzner.PickLoginKey()
		if err != nil {
			return err
		}
	}
	identityPath, err := expandUser(opts.LoginKeyPath)
	if err != nil {
		return err
	}
	if _, err := os.Stat(identityPath); err != nil {
		return run.Errorf("--login-key-path %s does not exist", identityPath)
	}

	var resolvedOpsDir string
	if opts.OpsDir != "" {
		resolvedOpsDir, err = expandUser(opts.OpsDir)
		if err != nil {
			return err
		}
		info, err := os.Stat(resolvedOpsDir)
		if err != nil || !info.IsDir() {
			return run.Errorf("--ops-dir %s isn't a directory", resolvedOpsDir)
		}
	}

	var branchPtr *string
	if opts.Branch != "" {
		branchPtr = &opts.Branch
	}

	fmt.Printf("[1/7] checking %s matches '%s' on Hetzner ...\n", identityPath, resolvedLoginKey)
	if err := hetzner.VerifyLoginKey(resolvedLoginKey, identityPath); err != nil {
		return err
	}

	fmt.Printf("[2/7] creating %s in %s (%s) ...\n", opts.ServerType, opts.Location, instanceName)
	scriptFile, err := os.CreateTemp("", "hcode-provision-*.sh")
	if err != nil {
		return err
	}
	scriptPath := scriptFile.Name()
	if _, err := scriptFile.WriteString(provision.BaseProvisionScript); err != nil {
		scriptFile.Close()
		os.Remove(scriptPath)
		return err
	}
	scriptFile.Close()
	serverID, ip, err := hetzner.CreateServer(
		instanceName, opts.ServerType, opts.Location, resolvedLoginKey, scriptPath,
		map[string]string{"hcode-instance": instanceName},
	)
	os.Remove(scriptPath)
	if err != nil {
		return err
	}

	fmt.Printf("[3/7] waiting for %s to finish booting (this installs Docker/Node/uv, ~1-2 min) ...\n", ip)
	if err := sshutil.WaitForSSH(ip, identityPath, 0); err != nil {
		return err
	}

	fmt.Println("[4/7] attaching the repo ...")
	repoState, err := attach.AttachRepo(ip, identityPath, instanceName, opts.RepoURL, branchPtr, opts.EnvFiles)
	if err != nil {
		return err
	}

	if opts.Worktrees > 0 {
		fmt.Printf("[5/7] adding %d worktree(s) for parallel lanes ...\n", opts.Worktrees)
		mainDest := fmt.Sprintf("%s/%s", config.RemoteCodeDir, repo.Name)
		for n := 2; n < opts.Worktrees+2; n++ {
			label := fmt.Sprintf("cc%d", n)
			worktreeDest := fmt.Sprintf("%s/%s-%s", config.RemoteCodeDir, repo.Name, label)
			cmd := provision.AddWorktreeCommand(mainDest, worktreeDest, fmt.Sprintf("%s/base", label))
			if _, err := sshutil.RunRemote(ip, cmd, identityPath); err != nil {
				return err
			}
			repoState.Worktrees = append(repoState.Worktrees, label)
			fmt.Printf("  %s  (branch %s/base)\n", worktreeDest, label)

			for _, local := range opts.EnvFiles {
				remotePath := fmt.Sprintf("%s/%s", worktreeDest, filepath.ToSlash(local))
				if err := sshutil.CopyTo(ip, local, remotePath, identityPath); err != nil {
					return err
				}
				fmt.Printf("    + %s (worktrees don't inherit untracked files on their own)\n", local)
			}

			if opts.PostWorktree != "" {
				fmt.Printf("    running %s for %s ...\n", opts.PostWorktree, label)
				cmd := fmt.Sprintf("cd %s && HCODE_WORKTREE_LABEL=%s bash %s", worktreeDest, label, opts.PostWorktree)
				if err := sshutil.RunRemoteStreaming(ip, cmd, identityPath); err != nil {
					return err
				}
			}
		}
	} else {
		fmt.Println("[5/7] no --worktrees requested, skipping")
	}

	var remoteOpsDir *string
	if resolvedOpsDir != "" {
		remote := fmt.Sprintf("%s/%s_ops", config.RemoteCodeDir, repo.Name)
		fmt.Printf("[6/7] copying %s -> %s ...\n", resolvedOpsDir, remote)
		if err := sshutil.CopyDirTo(ip, resolvedOpsDir, remote, identityPath); err != nil {
			return err
		}
		remoteOpsDir = &remote
	} else {
		fmt.Println("[6/7] no --ops-dir given, skipping")
	}

	if opts.PostClone != "" {
		mainDest := fmt.Sprintf("%s/%s", config.RemoteCodeDir, repo.Name)
		fmt.Printf("[7/7] running %s (repo-defined setup — output below) ...\n", opts.PostClone)
		cmd := fmt.Sprintf("cd %s && bash %s", mainDest, opts.PostClone)
		if err := sshutil.RunRemoteStreaming(ip, cmd, identityPath); err != nil {
			return err
		}
	} else {
		fmt.Println("[7/7] no --post-clone given, skipping")
	}

	var opsDirLocal *string
	if resolvedOpsDir != "" {
		opsDirLocal = &resolvedOpsDir
	}
	instance := state.Instance{
		Name:         instanceName,
		ServerID:     serverID,
		IP:           ip,
		Type:         opts.ServerType,
		Location:     opts.Location,
		LoginKey:     resolvedLoginKey,
		LoginKeyPath: identityPath,
		CreatedAt:    time.Now().UTC().Format(time.RFC3339Nano),
		Repos:        []state.Repo{repoState},
		OpsDir:       remoteOpsDir,
		OpsDirLocal:  opsDirLocal,
	}
	if err := state.Save(instance); err != nil {
		return err
	}

	fmt.Printf("\n%s is up at %s\n", instanceName, ip)
	fmt.Printf("  ssh:    hcode ssh %s\n", instanceName)
	if opts.Worktrees > 0 {
		fmt.Printf("  ssh:    hcode ssh %s --worktree cc2  (etc.)\n", instanceName)
	}
	fmt.Printf("  add:    hcode add <repo-url> --instance %s\n", instanceName)
	fmt.Printf("  kill:   hcode destroy %s\n", instanceName)

	if !opts.NoAttach {
		normalized := make([]string, 0, len(opts.Forwards))
		for _, f := range opts.Forwards {
			nf, err := forwards.NormalizeForward(f)
			if err != nil {
				return err
			}
			normalized = append(normalized, nf)
		}
		sshutil.Interactive(ip, identityPath, fmt.Sprintf("%s/%s", config.RemoteCodeDir, repo.Name), normalized)
	}
	return nil
}
